File: 001376_TimeNeededToInformAllEmployees/main.go
package main

import "fmt"

func numOfMinutes(n int, headID int, manager []int, informTime []int) int {
	direct_managers := make(map[int][]int, n)
	for i := 0; i < n; i++ {
		arr, has := direct_managers[manager[i]]
		if !has {
			arr = make([]int, 0)
		}
		direct_managers[manager[i]] = append(arr, i)
	}

	cost := -1

	q := make([][]int, 0)
	q = append(q, []int{headID, informTime[headID]})

	for len(q) > 0 {
		t := q[0]

		cost_ := t[1]
		if cost_ > cost {
			cost = cost_
		}

		direct_manager := t[0]
		for _, sub := range direct_managers[direct_manager] {
			q = append(q, []int{sub, cost_ + informTime[sub]})
		}

		q = q[1:]
	}

	return cost
}

func main() {
	fmt.Println(numOfMinutes(1, 0, []int{-1}, []int{0}))
	fmt.Println(numOfMinutes(6, 2, []int{2, 2, -1, 2, 2, 2}, []int{0, 0, 1, 0, 0, 0}))
	fmt.Println(numOfMinutes(8, 0, []int{-1, 5, 0, 6, 7, 0, 0, 0}, []int{89, 0, 0, 0, 0, 523, 241, 519}))
}

File: 001376_TimeNeededToInformAllEmployees/main.py
from typing import List
from collections import defaultdict, deque


class Solution:
    def numOfMinutes(self, n: int, headID: int, manager: List[int], informTime: List[int]) -> int:
        graph = defaultdict(list)
        for i in range(n):
            graph[manager[i]].append((i, informTime[manager[i]]))
            
        ret = 0
        queue = deque([(headID, 0)])
        while queue:
            cur_node, cur_t = queue.popleft()
            ret = max(ret, cur_t)
            for neighbor, dt in graph[cur_node]:
                queue.append((neighbor, cur_t+dt))
        return ret


if __name__ == "__main__":
    sln = Solution()
    sln.numOfMinutes(8, 0, [-1, 5, 0, 6, 7, 0, 0, 0], [89, 0, 0, 0, 0, 523, 241, 519])